pkg/operation/botanist/secrets.go

```diff
@@ -15,838 +15,189 @@
+		if err := b.deleteBasicAuthDependantSecrets(ctx, &gardenerResourceDataList); err != nil {
 			return err
 		}
 	}
 
-	// The kubeconfig secrets for the dependency-watchdog probes have now stopped using token to avoid the need for reconciliation
-	// on rotation. The old secrets containing tokens must be removed so that new secrets without token can be generated.
-	for _, secretName := range []string{common.DependencyWatchdogInternalProbeSecretName, common.DependencyWatchdogExternalProbeSecretName} {
-		secret := &corev1.Secret{}
-		if err := b.K8sSeedClient.Client().Get(ctx, kutil.Key(b.Shoot.SeedNamespace, secretName), secret); err != nil {
-			if apierrors.IsNotFound(err) {
-				continue
-			}
+	secretsGenerator := controlplanesecrets.NewSecretsGenerator(
+		gardenerResourceDataList,
+		basicAuthSecretAPIServer,
+		staticTokenConfig,
+		wantedCertificateAuthorities,
+		b.generateWantedSecretConfigs,
+	)
+
+	if shootWantsBasicAuth {
+		if err := secretsGenerator.GenerateAPIServerBasicAuth(); err != nil {
 			return err
 		}
-		if _, ok := secret.Data[secrets.DataKeyToken]; ok {
-			// The kubeconfig uses bearer token. Delete it to regenerate the new kubeconfig without bearer token.
-			if err := b.K8sSeedClient.Client().Delete(ctx, secret); client.IgnoreNotFound(err) != nil {
-				return err
-			}
-		}
-
-		// The kubeconfig does not use bearer token. No change required.
 	}
 
-	existingSecretsMap, err := b.fetchExistingSecrets(ctx)
-	if err != nil {
+	if err := secretsGenerator.GenerateStaticToken(); err != nil {
 		return err
 	}
 
-	certificateAuthorities, err := b.generateCertificateAuthorities(existingSecretsMap)
-	if err != nil {
+	if err := secretsGenerator.GenerateCertificateAuthorities(); err != nil {
 		return err
 	}
 
-	var basicAuthAPIServer *secrets.BasicAuth
-	if gardencorev1beta1helper.ShootWantsBasicAuthentication(b.Shoot.Info) {
-		basicAuthAPIServer, err = b.generateBasicAuthAPIServer(ctx, existingSecretsMap)
-		if err != nil {
-			return err
-		}
-	}
-
-	staticToken, err := b.generateStaticToken(ctx, existingSecretsMap)
-	if err != nil {
+	if err := secretsGenerator.GenerateSecrets(); err != nil {
 		return err
 	}
 
-	if err := b.deployOpenVPNTLSAuthSecret(ctx, existingSecretsMap); err != nil {
+	shootState := &gardencorev1alpha1.ShootState{ObjectMeta: kutil.ObjectMeta(b.Shoot.Info.Namespace, b.Shoot.Info.Name)}
+	if _, err := controllerutil.CreateOrUpdate(ctx, b.K8sGardenClient.Client(), shootState, func() error {
+		shootState.Spec.Gardener = secretsGenerator.GardenerResourceDataList
+		return nil
+	}); err != nil {
 		return err
 	}
 
-	wantedSecretsList, err := b.generateWantedSecrets(basicAuthAPIServer, staticToken, certificateAuthorities)
+	b.ShootState = shootState
+
+	return nil
+}
+
+// DeploySecrets takes all existing secrets from the ShootState resource and deploys them in the shoot's control plane.
+func (b *Botanist) DeploySecrets(ctx context.Context) error {
```

Comment:
Is my understanding correct that this function reads the `ShootState` again after it was previously generated by `GenerateAndSaveSecrets`? If yes, can we avoid these network calls by only reading the `ShootState` once and then passing it internally to the next function?

Reply:
It reads the saved struct from the botanist. So what you suggested is already happening :)